framework/src/main/java/org/checkerframework/common/returnsreceiver/ReturnsReceiverAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,90 @@
+import org.checkerframework.common.returnsreceiver.qual.BottomThis;
+import org.checkerframework.common.returnsreceiver.qual.This;
+import org.checkerframework.common.returnsreceiver.qual.UnknownThis;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.typeannotator.ListTypeAnnotator;
+import org.checkerframework.framework.type.typeannotator.TypeAnnotator;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/** The type factory for the Returns Receiver Checker. */
+public class ReturnsReceiverAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    /** The {@code @}{@link This} annotation. */
+    final AnnotationMirror THIS_ANNOTATION;
+
+    /**
+     * Create a new {@code ReturnsReceiverAnnotatedTypeFactory}.
+     *
+     * @param checker the type-checker associated with this factory
+     */
+    public ReturnsReceiverAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        THIS_ANNOTATION = AnnotationBuilder.fromClass(elements, This.class);
+        this.postInit();
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return getBundledTypeQualifiers(BottomThis.class, UnknownThis.class, This.class);
+    }
+
+    @Override
+    protected TypeAnnotator createTypeAnnotator() {
+        return new ListTypeAnnotator(
+                new ReturnsReceiverTypeAnnotator(this), super.createTypeAnnotator());
+    }
+
+    /** A TypeAnnotator to add the {@code @}{@link This} annotation. */
+    private class ReturnsReceiverTypeAnnotator extends TypeAnnotator {
+
+        /**
+         * Create a new ReturnsReceiverTypeAnnotator.
+         *
+         * @param typeFactory the {@link AnnotatedTypeFactory} associated with this {@link
+         *     TypeAnnotator}
+         */
+        public ReturnsReceiverTypeAnnotator(AnnotatedTypeFactory typeFactory) {
+            super(typeFactory);
+        }
+
+        @Override
+        public Void visitExecutable(AnnotatedTypeMirror.AnnotatedExecutableType t, Void p) {
+
+            // skip constructors, as we never need to add annotations to them
+            if (t.getElement().getKind() == ElementKind.CONSTRUCTOR) {
+                return super.visitExecutable(t, p);
+            }
+
+            AnnotatedTypeMirror returnType = t.getReturnType();
+
+            // If any FluentAPIGenerator indicates the method returns this,
+            // add an @This annotation on the return type.
+            if (FluentAPIGenerator.check(t)) {
+                if (!returnType.isAnnotatedInHierarchy(THIS_ANNOTATION)) {
+                    returnType.addAnnotation(THIS_ANNOTATION);
+                }
+            }
+
+            // If return type is annotated with @This, add @This annotation
```

Comment:
https://github.com/typetools/checker-framework/pull/3199#discussion_r410451990
Did you try using `@DefaultFor({TypeUseLocation.RECEIVER})` on the declaration of `@This`?

Reply:
I tried adding this to the declaration of ```@This```, but below tests fails for this change:

- returnsreceiver
- all-systems/java8/memberref
- all-systems/java8/lambda
- all-systems/java8inference

Is there anything else I should do to make this change or we should keep it as it is ?

Reply:
It seems the primary issue here is with method references.  Here is one false-positive report we get:
```
tests/all-systems/java8/memberref/MemberReferences.java:119: error: [methodref.receiver.bound.invalid] Incompatible receiver type
        Consumer<Number> f1 = bound::func1;
                              ^
  Method
    @UnknownThis BoundWithArg<@UnknownThis Number>
  is not a valid method reference for
    void func1(@This BoundWithArg<@UnknownThis Number> this, @UnknownThis Number p0) in BoundWithArg
  found   : @UnknownThis BoundWithArg<@UnknownThis Number>
  required: @This BoundWithArg<@UnknownThis Number>
```
@wmdietl I'm not sure why this is an error; seems like an interaction between polymorphic annotations and generics?  Maybe it's a Checker Framework bug?

Reply:
It looks like you don't have a lot of lambda/method reference tests and running on all-systems with the change revealed that...

The following test without generics gives a similar error (incorrectly I named it Lambda, but then I used a method reference... I'm sure you can write something similar that actually uses a lambda):

```java
import org.checkerframework.common.returnsreceiver.qual.*;

class Lambda {

    @This Lambda set(Object o) {
        return this;
    }

    interface Setter {
        @This Object consume(Object p);
    }

    Setter co = this::set;
}
```

gives:

```
Lambda.java:13: error: [methodref.receiver.bound.invalid] Incompatible receiver type
    Setter co = this::set;
                ^
  Method
    @UnknownThis Lambda
  is not a valid method reference for
    @This Lambda set(@This Lambda this, @UnknownThis Object p0) in Lambda
  found   : @UnknownThis Lambda
  required: @This Lambda
1 error
```

This is with tip of this branch, so independent of the proposed change.
Should that be an error? What is the allowed interaction between `@This` and lamdas/method references?

Reply:
Good catch!  I am not sure what is going on here.  Our checker should be adding `@This` on the receiver of `set()`.  @wmdietl do you happen to know why this error message says `@UnknownThis Lambda` as opposed to `@This Lambda`?

Reply:
No, I don't. Maybe there is a problem with how polymorphic annotations are resolved on receivers?

Reply:
The error message seems odd, as it says `Method @UnknownThis Lambda`, which is just a declared type, not a method type.

Reply:
I think the error message is just phrased poorly.  What seems to be going on is that there is a subtype check that fails because `@UnknownThis Lambda` is not a subtype of `@This Lambda`.  I've added this example as a test case now but renamed it to `MethodRef`.

I am not too sure about how to fix this one, and whether it's a bug in our checker or not.  @smillst any chance you could take a look?

Reply:
Thinking more, this example is actually quite subtle, and I'm now not sure if there should be a type error or not.  Consider this extended version of the example:
```java
import org.checkerframework.common.returnsreceiver.qual.*;

class MethodRef {

    @This MethodRef set(Object o) {
        return this;
    }

    interface Setter {
        @This Object consume(Object p);
    }

    Setter co = this::set;

    public static void main(String[] args) {
        MethodRef m = new MethodRef();
        // this assertion fails!
        assert m.co.consume(new Object()) == m.co : "not returning receiver";
    }
}
```
I read the `@This` annotation on the `Setter.consume()` return type as indicating that any implementation of `Setter.consume()` should return its receiver.  But if we allow the above code to type check, that property does not hold (since `this::set` will return the value of `this` bound by the method reference, not the receiver passed into the call to `consume()`).

So maybe there _should_ be an error in this case (perhaps with a better message)?  @wmdietl @kelloggm @mernst thoughts?

Reply:
The Checker Framework does not handle polymorphic qualifiers receivers on the functional methods to member references are assigned.  For example, the receiver of `Setter#consume` has a polymorphic annotation on its receiver so, `Setter co = this::set;` causes a weird error.  There should be some sort of error, but `methodref.receiver.bound.invalid` seems to be the wrong one. 

We have one open issue on the https://github.com/typetools/checker-framework/issues/2931.  I've added an example to that issue using the Tainting Checker that shows this problem.

I would just work around this problem in this checker, which I think that's why you wrote the defaulting in place to begin with.  So, maybe just add a comment pointing to #2931?

Reply:
Thanks a lot, @smillst!  It seems like we will run into this issue whether `@This` is the default for receivers or not.  So I think ideally we should follow @wmdietl's suggestion to use `@DefaultFor({TypeUseLocation.RECEIVER})` (which will save us a bit of code).  But then, some of the `all-systems` tests will fail.  Is there a way to suppress some warnings but still use the `all-systems` tests?

Reply:
You can add 
```
@SuppressWarnings("returnsreceiver: methodref.receiver.bound.invalid") // See Issue #2931
```
to any `all-systems` tests that need it.  In general, it is acceptable to suppress a type system specific warning in the all-system test so long as an explanation is given.

Reply:
Unfortunately, it seems that currently, we get too many false positives with method references if we use `@DefaultFor({TypeUseLocation.RECEIVER})` on `@This`.  E.g., with receiver defaulting, we get an error in this variant:
```java
class MethodRef {
    void doNothing() {}

    interface Fun {
        void run();
    }

    Fun f = this::doNothing;
}
```
I think _any_ method reference that binds the receiver will yield an error with the returns receiver checker if receivers are defaulted to `@This`.  I think this is an unacceptable level of false positives, so for now we will not do the defaulting but will leave a comment as to why we don't do it.